questionnaire/questionnaire-service.js

```diff
@@ -588,18 +586,131 @@ function createQuestionnaireService({
+        // 1) get questionnaire instance.
+        const questionnaire = await getQuestionnaire(questionnaireId);
+        if (!questionnaire) {
+            throw new VError(
+                {
+                    name: 'ResourceNotFound'
+                },
+                `Questionnaire with questionnaireId "${questionnaireId}" does not exist`
+            );
+        }
+
+        // 2) get questionnaire instance's submission status.
+        const submissionStatus = await getQuestionnaireSubmissionStatus(questionnaireId);
+        // 3) are we currently, or have we been on this questionnaire's summary page?
+        // we infer a questionnaire is complete if the user has visited the summary page.
+        const isQuestionnaireComplete = questionnaire.progress.includes(
+            questionnaire.routes.summary
+        );
+
+        if (!isQuestionnaireComplete) {
+            throw new VError(
+                {
+                    name: 'ResourceConflict'
+                },
+                `Questionnaire with ID "${questionnaireId}" is not in a submittable state`
+            );
+        }
+
+        // if the submission status is anything other than 'NOT_STARTED' then it
+        // means that the submission resource has been previously created.
+        // also skip over this for failed application so they can be resubmitted.
+        if (!['NOT_STARTED', 'FAILED'].includes(submissionStatus)) {
+            throw new VError(
+                {
+                    name: 'ResourceConflict'
+                },
+                `Submission resource with ID "${questionnaireId}" already exists`
+            );
+        }
+
+        // check all answers are correct.
+        await validateAllAnswers(questionnaireId);
+
+        // TODO: refactor `getSubmissionResource` to be more intuitive.
+        const resourceCollection = await getSubmissionResource(questionnaireId, true);
+
+        createAnswers(questionnaireId, questionnaire.routes.summary, {});
+
+        return resourceCollection;
+    }
+
+    async function getSubmissions(submissionStatus) {
+        const status = submissionStatus.toUpperCase();
+        const questionnaireIds = await db.getQuestionnaireIdsBySubmissionStatus(status);
+
+        const promises = questionnaireIds.map(async questionnaireId => {
+            const submissionResource = await getSubmissionResource(questionnaireId);
+            return submissionResource;
+        });
+        const resourceCollection = await Promise.all(promises);
+        return resourceCollection;
+    }
+
+    async function postSubmissions(submissionStatus) {
+        const status = submissionStatus.toUpperCase();
+        let questionnaireIds;
+        try {
+            questionnaireIds = await db.getQuestionnaireIdsBySubmissionStatus(status);
+        } catch (err) {
+            // no IDs means there is nothing to submit, so return early.
+            return [];
+        }
```

Comment:
If it's already thrown in the `getQuestionnaireIdsBySubmissionStatus` function, I don't know that this will ever catch anything? If it does though, an empty array doesn't seem very helpful.